File: apps/api/app/schemas/ai_writing.py
"""AI Writing schemas."""

from pydantic import BaseModel, Field

from app.schemas.rag_chat import GroundedPassage


class AutocompleteRequest(BaseModel):
    prefix_text: str = Field(max_length=32000)
    suffix_text: str | None = None
    paragraph_context: str = ""
    section_heading: str | None = None
    mode: str = "ghost"  # 'ghost' | 'continuation'
    paper_ids: list[str] | None = None


class AutocompleteResponse(BaseModel):
    text: str
    grounding_state: str  # 'source-grounded' | 'ai-inference' | 'general-knowledge'
    source_passages: list[GroundedPassage] = Field(default_factory=list)
    mode: str = "ghost"
    latency_ms: int | None = None


class AutocompleteSettings(BaseModel):
    enabled: bool = True
    engine: str = "auto"  # 'auto' | 'tabby' | 'cloud' | 'ollama'
    base_url: str | None = None
    model: str | None = None


class AutocompleteSettingsUpdate(BaseModel):
    enabled: bool | None = None
    engine: str | None = None
    base_url: str | None = None
    model: str | None = None


class AutocompleteSettingsResponse(AutocompleteSettings):
    pass


class AutocompleteProbeResponse(BaseModel):
    reachable: bool
    base_url: str | None = None


class AIEditRequest(BaseModel):
    text: str = Field(max_length=32000)
    action: str  # 'clarity' | 'academic' | 'simplify' | 'shorten' | 'expand' | 'grammar' | 'flow' | 'translate' | 'explain'
    target_language: str | None = "English"
    paragraph_context: str | None = None
    surrounding_context: str | None = None
    paper_ids: list[str] | None = None


class AIEditResponse(BaseModel):
    original_text: str
    suggested_text: str
    action: str
    explanation: str | None = None
    grounding_state: str = "general-knowledge"
    changes_summary: str | None = None
    sources: list[GroundedPassage] = Field(default_factory=list)
    latency_ms: int | None = None


class AIOutlineSection(BaseModel):
    id: str
    title: str
    level: int = 1
    description: str | None = None
    key_points: list[str] = Field(default_factory=list)
    suggested_passages: list[GroundedPassage] = Field(default_factory=list)


class AIOutlineRequest(BaseModel):
    topic: str
    research_question: str | None = None
    paper_ids: list[str] | None = None
    target_sections_count: int | None = 6
    context_notes: str | None = None


class AIOutlineResponse(BaseModel):
    topic: str
    research_question: str | None = None
    sections: list[AIOutlineSection] = Field(default_factory=list)
    estimated_word_count: int = 4000
    grounding_state: str = "general-knowledge"
    sources: list[GroundedPassage] = Field(default_factory=list)
    latency_ms: int | None = None

File: apps/web/src/lib/api/aiWriting.ts
import { request } from './client';
import type { GroundingState, GroundedPassage, AIEditActionType } from '@openresearch/ai';

export interface AICompletionResponseDTO {
  text: string;
  grounding_state: GroundingState;
  source_passages: GroundedPassage[];
  mode: string;
  latency_ms: number;
}

export interface AIEditResponseDTO {
  original_text: string;
  suggested_text: string;
  action: AIEditActionType | string;
  explanation?: string;
  changes_summary?: string;
  grounding_state: GroundingState;
  sources: GroundedPassage[];
  latency_ms: number;
}

export interface AIOutlineSectionDTO {
  id: string;
  title: string;
  level: number;
  description?: string;
  key_points: string[];
  suggested_passages: GroundedPassage[];
}

export interface AIOutlineResponseDTO {
  topic: string;
  research_question?: string;
  sections: AIOutlineSectionDTO[];
  estimated_word_count: number;
  grounding_state: GroundingState;
  sources: GroundedPassage[];
  latency_ms: number;
}

export const aiWritingApi = {
  autocomplete: (
    projectId: string,
    data: {
      prefix_text: string;
      suffix_text?: string;
      paragraph_context?: string;
      section_heading?: string;
      mode?: 'ghost' | 'continuation';
      paper_ids?: string[];
    }
  ) =>
    request<AICompletionResponseDTO>(`/projects/${projectId}/ai/autocomplete`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  edit: (
    projectId: string,
    data: {
      text: string;
      action: AIEditActionType | string;
      target_language?: string;
      paragraph_context?: string;
      surrounding_context?: string;
      paper_ids?: string[];
    }
  ) =>
    request<AIEditResponseDTO>(`/projects/${projectId}/ai/edit`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
  outline: (
    projectId: string,
    data: {
      topic: string;
      research_question?: string;
      paper_ids?: string[];
      target_sections_count?: number;
      context_notes?: string;
    }
  ) =>
    request<AIOutlineResponseDTO>(`/projects/${projectId}/ai/outline`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),
};
